libftl/ingest.c

```diff
@@ -1,22 +1,150 @@
+
+typedef struct {
+    ftl_ingest_t *ingest;
+    ftl_stream_configuration_private_t *ftl;
+}_tmp_ingest_thread_data_t;
+
+static int _ping_server(const char *ip, int port) {
+
+  SOCKET sock;
+  struct sockaddr_in server_addr;
+  unsigned char buf[sizeof(struct in_addr)];
+  uint8_t dummy[4];
+  struct timeval start, stop, delta;
+  int retval = -1;
+
+  do {
+    if ((sock = socket(AF_INET, SOCK_DGRAM, 0)) == INVALID_SOCKET)
+    {
+      break;
+    }
+
+    if (inet_pton(AF_INET, ip, buf) == 0) {
+      break;
+    }
+
+    //Prepare the sockaddr_in structure
+    server_addr.sin_family = AF_INET;
+    memcpy((char *)&server_addr.sin_addr.s_addr, (char *)buf, sizeof(buf));
+    server_addr.sin_port = htons(port);
+
+    set_socket_recv_timeout(sock, 500);
+
+    gettimeofday(&start, NULL);
+
+    if (sendto(sock, dummy, sizeof(dummy), 0, (struct sockaddr*) &server_addr, sizeof(struct sockaddr_in)) == SOCKET_ERROR) {
+      break;
+    }
+
+    if (recv(sock, dummy, sizeof(dummy), 0) < 0) {
+      break;
+    }
+
+    gettimeofday(&stop, NULL);
+    timeval_subtract(&delta, &stop, &start);
+    retval = (int)timeval_to_ms(&delta);
+  } while (0);
+
+  shutdown_socket(sock, SD_BOTH);
+  close_socket(sock);
+
+  return retval;
+}
+
+OS_THREAD_ROUTINE _ingest_get_rtt(void *data) {
+    _tmp_ingest_thread_data_t *thread_data = (_tmp_ingest_thread_data_t *)data;
+    ftl_stream_configuration_private_t *ftl = thread_data->ftl;
+    ftl_ingest_t *ingest = thread_data->ingest;
+    int ping;
+
+    ingest->rtt = 1000;
+
+    if ((ping = _ping_server(ingest->ip, INGEST_PING_PORT)) >= 0) {
+        ingest->rtt = ping;
+    }
+
+    return 0;
+}
+
+ftl_status_t ftl_find_closest_ingest(const char* ingestIps[], const char* ingestNames[], int ingestsCount, char* bestIngestIpComputed)
+{
+    ftl_ingest_t* ingestElement;
+    int i;
+
+    if ((ingestElement = malloc(sizeof(ftl_ingest_t) * ingestsCount)) == NULL) {
+        return FTL_MALLOC_FAILURE;
+    }
+
+    for (i = 0; i < ingestsCount; i++)
+    {
+        strcpy_s(ingestElement[i].name, sizeof(ingestElement[i].name), ingestNames[i]);
+        strcpy_s(ingestElement[i].ip, sizeof(ingestElement[i].ip), ingestIps[i]);
+        ingestElement[i].rtt = 500;
+        ingestElement[i].next = NULL;
+    }
+
+    OS_THREAD_HANDLE *handle;
+    _tmp_ingest_thread_data_t *data;
+    ftl_ingest_t *best = NULL;
+
+    if ((handle = (OS_THREAD_HANDLE *)malloc(sizeof(OS_THREAD_HANDLE) * ingestsCount)) == NULL) {
+        return FTL_MALLOC_FAILURE;
+    }
+
+    if ((data = (_tmp_ingest_thread_data_t *)malloc(sizeof(_tmp_ingest_thread_data_t) * ingestsCount)) == NULL) {
+        return FTL_MALLOC_FAILURE;
+    }
+
+    /*query all the ingests about cpu and rtt*/
+    for (i = 0; i < ingestsCount; i++) {
+        handle[i] = 0;
+        data[i].ingest = &ingestElement[i];
+        data[i].ftl = NULL; // This is not used by _ingest_get_rtt
+        os_create_thread(&handle[i], NULL, _ingest_get_rtt, &data[i]);

```

Comment:
Using the same logic as _ingest_get_hosts. I think its ok. Even if one of the threads fail, you can still recover. Although a thread creation failure probably means something much more malign.